### PR TITLE
gh-124153: Fix unstable optimization of `PyType_GetBaseByToken()` and friends

### PR DESCRIPTION
**Pull request by neonene**
An alternative PR to #124323.

This helps MSVC to optimize `PyType_GetBaseByToken()` and `PyType_GetModuleByDef()` on PGO builds without current workaround.

cc @encukou

* Issue: gh-124153

### CODE REVIEW

File: Objects/typeobject.c [hunk 1]
```diff
@@ -5297,12 +5297,12 @@ get_base_by_token_recursive(PyTypeObject *type, void *token)
     Py_ssize_t n = PyTuple_GET_SIZE(bases);
     for (Py_ssize_t i = 0; i < n; i++) {
         PyTypeObject *base = _PyType_CAST(PyTuple_GET_ITEM(bases, i));
-        if (!_PyType_HasFeature(base, Py_TPFLAGS_HEAPTYPE)) {
-            continue;
-        }
-        if (((PyHeapTypeObject*)base)->ht_token == token) {
+        // The MRO version can slow down if this takes the same approach,
+        // which also affects the same conditions in the functions above.
+        if (PyType_GetSlot(base, Py_tp_token) == token) {
```
Comment (encukou) [inline Objects/typeobject.c]: I'd prefer reverting this change: in `typeobject.c`, use the type internals directly.

Or did this somehow improve performance?

Comment (neonene) [reply Objects/typeobject.c]: Do you mean the performance of this recursive function?  If so, I'll reconsider.

Comment (encukou) [reply Objects/typeobject.c]: I meant that as a question: did you do this change to improve performance?

**Comment by neonene** [comment]
One of suspects that exhausts MSVC:
```diff
 static PyTypeObject *
-get_base_by_token_recursive(PyTypeObject *type, void *token)
+get_base_by_token_recursive(PyObject *bases, void *token)
```

**Comment by neonene** [comment]
@encukou Please review again. This should not be bad for performance on average.

**Comment by neonene** [comment]
Bad example:

```c
    PyTypeObject *base = NULL;
    if (((PyHeapTypeObject*)type)->ht_token == token) {
found:
        if (result != NULL) {
            *result = (PyTypeObject *)Py_NewRef(base ? base : type);
        }
        return 1;
    }
```

Are you interested in removing the checker function like the current PR?  If not, I'll give up.

**Review by encukou — APPROVED** [review]
This looks good! Thank you, and apologies for the delay.

**Pull request merged**